qiskit/transpiler/passes/synthesis/qubit_tracker.py

```diff
@@ -0,0 +1,123 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A qubit state tracker for synthesizing operations with auxiliary qubits."""
+
+from __future__ import annotations
+from collections.abc import Iterable
+from dataclasses import dataclass
+
+
+@dataclass
+class QubitTracker:
+    """Track qubits per index as and their state.
+
+    The states are distinguished into clean (meaning in state :math:`|0\rangle`) or dirty (an
+    unkown state).
+    """
+
+    # This could in future be extended to track different state types, if necessary.
+    # However, using sets of integers here is much faster than e.g. storing a dictionary with
+    # {index: state} entries.
+    qubits: tuple[int]
+    clean: set[int]
+    dirty: set[int]
+
+    def num_clean(self, active_qubits: Iterable[int] | None = None):
+        """Return the number of clean qubits, not considering the active qubits."""
+        # this could be cached if getting the set length becomes a performance bottleneck
+        return len(self.clean.difference(active_qubits or set()))
+
+    def num_dirty(self, active_qubits: Iterable[int] | None = None):
+        """Return the number of dirty qubits, not considering the active qubits."""
+        return len(self.dirty.difference(active_qubits or set()))
+
+    def borrow(self, num_qubits: int, active_qubits: Iterable[int] | None = None) -> list[int]:
+        """Get ``num_qubits`` qubits, excluding ``active_qubits``."""
+        if active_qubits is None:
+            active_qubits = []
+
+        if num_qubits > (available := len(self.qubits) - len(active_qubits)):
+            raise RuntimeError(f"Cannot borrow {num_qubits} qubits, only {available} available.")
+
+        return [qubit for qubit in self.qubits if qubit not in active_qubits][:num_qubits]
```

Comment:
We can optimized the `borrow` method by converting `active_qubits` to a set for efficient membership testing. This reduces the complexity of the `not in` checks within the list comprehension and simplifies the code. The updated method ensures faster execution, especially when `active_qubits` is large. 
```suggestion
        active_qubits = set(active_qubits or [])
        available_qubits = [qubit for qubit in self.qubits if qubit not in active_qubits]

        if num_qubits > len(available_qubits):
            raise RuntimeError(
                f"Cannot borrow {num_qubits} qubits, only {len(available_qubits)} available."
            )

        return available_qubits[:num_qubits]
```
The approach simplifies the check for available qubits, making it more readable and potentially more efficient.